MiniSoup (Independent Publisher)

DEmodoriGatsuO opened the pull request:
# MiniSoup HTML Parser (Independent Publisher)

## Overview
MiniSoup is a lightweight HTML parsing library for Power Automate. Inspired by Beautiful Soup, this connector enables direct HTML parsing, data extraction, and web analysis within Power Automate without requiring external services.

This connector is ideal for automated collection of web information, extraction of structured data, website monitoring, and other automation tasks requiring HTML parsing.

## Key Features
- **Lightweight HTML Parsing**: Efficiently extract data from HTML content
- **CSS Selector Support**: Target specific elements using familiar CSS selector syntax
- **XPath Support**: Alternative element targeting using XPath expressions
- **Table Parsing**: Extract structured data from HTML tables
- **Element Manipulation**: Find, filter, and extract information from HTML elements
- **Web Fetching**: Built-in capability to fetch HTML from URLs

## Use Case Examples
- **E-commerce Price Monitoring**: Regularly capture and analyze competitors' pricing information
- **News Headline Collection**: Retrieve the latest headlines from major news sites and post to Teams
- **Bulk Web Data Extraction**: Collect and analyze structured data from multiple pages
- **Table Data Retrieval**: Import tabular data from websites into Excel or databases
- **Metadata Extraction**: Retrieve OGP and meta tag information from websites

## Operations Overview
1. **fetch_html**: Retrieves HTML content from a specified URL
2. **select**: Selects HTML elements matching the provided selector
3. **extract**: Extracts specific attribute values from HTML elements matching the provided selector
4. **find_all**: Finds all HTML elements matching the specified tag name and optional attributes
5. **parse_table**: Parses an HTML table into structured data with headers and rows

## Technical Approach
MiniSoup adopts a lightweight approach designed to work effectively within the constraints of Power Automate custom connectors. It uses regex-based HTML parsing to analyze key HTML structures, extract elements, and manipulate attributes without relying on external libraries.

The design focuses on performance and memory efficiency, allowing it to operate efficiently even in environments with a 2-minute timeout limit.

## Screenshots
![image](https://github.com/user-attachments/assets/5bfa1f3b-546a-4d12-b6ec-62c5c6016129)

## Authentication Method
This connector does not require authentication.

## Disclaimer
This connector uses third-party websites and APIs. Please use it in accordance with the terms of service of third-party services. Users should be mindful of legal and ethical considerations related to web scraping.

## Developer Information
- Developer: Shogo Shindo
- Email: demodori.gatsuo@gmail.com
- GitHub: DEmodoriGatsuO

DEmodoriGatsuO committed:
Create intro.md

DEmodoriGatsuO commented:
@microsoft-github-policy-service agree

DEmodoriGatsuO commented:
@microsoft/connectors
I've agreed to the CLA and would like to know if there's any additional information needed for this proposal. Looking forward to next steps for the MiniSoup connector submission.

vmanoharas commented:
Hi, @DEmodoriGatsuO, 

Thank you for the proposal!

Kindly submit your artifacts and update the readme.md file as per the below template.  Additionally remove the proposal from the title.

DEmodoriGatsuO committed:
Add script.csx

Add script.csx

DEmodoriGatsuO committed:
Add files via upload

Add files via upload

DEmodoriGatsuO commented:
![image1](https://github.com/user-attachments/assets/8f800504-89ae-4414-b07c-80e5d84bc248)
![image2](https://github.com/user-attachments/assets/d47c98f3-e05e-4550-8035-4ec5cb28d346)
![image3](https://github.com/user-attachments/assets/f3b4a821-aa88-479c-883c-282c2708be16)
![image4](https://github.com/user-attachments/assets/17fa44d1-0353-487c-9a53-5a32ab5d1853)
![image5](https://github.com/user-attachments/assets/df620ab4-01a0-47c3-81d6-7707b883feb8)

### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 
If this is your first time submitting to GitHub and you need some help, please sign up for this [[session](https://forms.office.com/pages/responsepage.aspx?id=KtIy2vgLW0SOgZbwvQuRaXDXyCl9DkBHq4A2OG7uLpdUMTFJWFFGVUxBNUFZQjZWRUdaOE5BMFkwNS4u)](https://forms.office.com/pages/responsepage.aspx?id=KtIy2vgLW0SOgZbwvQuRaXDXyCl9DkBHq4A2OG7uLpdUMTFJWFFGVUxBNUFZQjZWRUdaOE5BMFkwNS4u). 
- [x] I attest that the connector doesn't exist on the Power Platform today. I've verified by checking the pull requests in GitHub and by searching for the connector on the platform or in the documentation. 
- [x] I attest that the connector works and I verified by deploying and testing all the operations. 
- [x] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [x] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [x] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [x] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [x] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [x] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [x] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [x] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md.

DEmodoriGatsuO commented:
Hi @vmanoharas,

Thank you for reviewing my proposal! I've prepared all the necessary artifacts for the MiniSoup HTML Parser connector and have updated the pull request as requested:

- Removed "Proposal -" from the title
- Added the required files:
  - apiDefinition.swagger.json
  - script.csx
  - apiProperties.json
  - readme.md (following the template)

The connector is a lightweight HTML parsing library inspired by Beautiful Soup, enabling HTML element analysis and data extraction directly within Power Automate flows without requiring external services.

I've tested all operations (FetchHTML, SelectElements, ExtractValues, FindAllElements, ParseTable) and validated the swagger file. Screenshots of test operations and successful flows have been added to the PR description.

Please let me know if you need any additional information or have questions about the implementation.

Thank you for your consideration!

vmanoharas commented:
Hi @DEmodoriGatsuO, 

Thank you for submitting artifacts for this connector. However, kindly remove intro.md and icon.png files. 

![image](https://github.com/user-attachments/assets/07d93ff6-3245-463d-9d18-b4661a44157d)

DEmodoriGatsuO committed:
Delete independent-publisher-connectors/MiniSoup/icon.png

DEmodoriGatsuO committed:
Delete independent-publisher-connectors/MiniSoup/intro.md

DEmodoriGatsuO commented:
Hi @vmanoharas

Thank you for your feedback! I'll remove the intro.md and icon.png files as requested.

I've now updated the pull request to include only the necessary files:
- apiDefinition.swagger.json
- script.csx
- apiProperties.json
- readme.md

All other files have been removed. Thank you for guiding me through this process.

vmanoharas reviewed (dismissed):
Dear Partner

I hope you are doing well.
Congratulations, your pull request is approved. We will move forward with the certification process. We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments that typically takes 3-4 weeks.
Please let us know if you have any questions.
Thank you very much for working with us.

DEmodoriGatsuO commented:
Dear Team,

Thank you very much for the approval notification. I'm delighted to hear that my pull request has been accepted and that the certification process will be moving forward.

I understand that the production deployment process typically takes 3-4 weeks, and I appreciate your team's efforts in preparing the connector for deployment.

Should I have any questions during this process, I'll be sure to reach out. Thank you again for your support and collaboration.

Best regards,
Shogo Shindo

vmanoharas commented:
Hello @DEmodoriGatsuO, 

Unfortunately, during certification process your connector caught in the below swagger validation errors.  Kindly resolve these errors. 

##[error]src\codelessPreview\MiniSoupHTMLParser\apiDefinition.swagger.json(,): error ExtensionNameNotValid: The property 'scriptOperations' is not valid. If this property is an extension, its name must begin with "x-". ValuePath: 'scriptOperations' Value: 'scriptOperations'

##[error]src\codelessPreview\MiniSoupHTMLParser\apiDefinition.swagger.json(,): error DefaultResponseHasSchema: The 'default' response should not have schema definition. Schemas should be defined on expected responses only. ValuePath: 'paths/~1fetch-html/post/responses/default' Value: ''

##[error]src\codelessPreview\MiniSoupHTMLParser\apiDefinition.swagger.json(,): error DefaultResponseHasSchema: The 'default' response should not have schema definition. Schemas should be defined on expected responses only. ValuePath: 'paths/~1select/post/responses/default' Value: ''

##[error]src\codelessPreview\MiniSoupHTMLParser\apiDefinition.swagger.json(,): error DefaultResponseHasSchema: The 'default' response should not have schema definition. Schemas should be defined on expected responses only. ValuePath: 'paths/~1extract/post/responses/default' Value: ''

##[error]src\codelessPreview\MiniSoupHTMLParser\apiDefinition.swagger.json(,): error DefaultResponseHasSchema: The 'default' response should not have schema definition. Schemas should be defined on expected responses only. ValuePath: 'paths/~1find-all/post/responses/default' Value: ''

##[error]src\codelessPreview\MiniSoupHTMLParser\apiDefinition.swagger.json(,): error DefaultResponseHasSchema: The 'default' response should not have schema definition. Schemas should be defined on expected responses only. ValuePath: 'paths/~1parse-table/post/responses/default' Value: ''

##[error]src\codelessPreview\MiniSoupHTMLParser\apiDefinition.swagger.json(,): error RequiredPropertyMissing: The 'contact' property is required. ValuePath: 'info' Value: 'contact'

##[error]src\codelessPreview\MiniSoupHTMLParser\apiDefinition.swagger.json(,): error InvalidConnectorCategory: The connector category is invalid. Available categories are: [AI, Business Management, Business Intelligence, Collaboration, Commerce, Communication, Content and Files, Finance, Data, Human Resources, Internet of Things, IT Operations, Lifestyle and Entertainment, Marketing, Productivity, Sales and CRM, Security, Social Media, Website]. ValuePath: 'x-ms-connector-metadata/2/propertyValue' Value: 'Data;Web'

DEmodoriGatsuO committed:
fix: resolve swagger validation errors in API definition

## Swagger Validation Errors Fixed

This commit addresses all validation errors identified during the Power Automate connector certification process:

1. Changed 'scriptOperations' to 'x-ms-scriptOperations' to comply with extension naming conventions
2. Removed schema definitions from all 'default' responses
3. Added required 'contact' information to the info section
4. Corrected connector category from 'Data;Web' to 'Data;Website'

All changes have been validated against Power Automate connector requirements.

DEmodoriGatsuO commented:
Hello @vmanoharas,

Thank you for your detailed feedback regarding our MiniSoup HTML Parser connector submission. We have addressed all the validation errors you identified in your report:

1. Extension Naming Convention:
   - Fixed by changing 'scriptOperations' to 'x-ms-scriptOperations' to comply with the standard prefix requirement for extension properties

2. Default Response Schema:
   - Removed schema definitions from all 'default' responses across all endpoints
   - Maintained only the description field as required

3. Contact Information:
   - Added the required 'contact' information to the info section with support name, URL, and email address

4. Connector Categories:
   - Updated the connector category from 'Data;Web' to 'Data;Website' to align with the approved category list

Please let us know if you need any further modifications or have any questions.

Thank you for your assistance in this certification process.

Best regards,
Shogo Shindo

vmanoharas commented:
Thank you @DEmodoriGatsuO, we will validate and get back to you if any further changes required.

vmanoharas commented on independent-publisher-connectors/MiniSoup/apiProperties.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "properties": {
+    "connectionParameters": {},
+    "iconBrandColor": "#e0e0e0",
```
Independent publishers: Set iconBrandColor to "#da3b01"

vmanoharas reviewed (commented):
Hi @DEmodoriGastuO

Thank you for fixing the swagger validation issues. 

Please change the iconBrand color to "#da3b01".

DEmodoriGatsuO committed:
Update iconBrandColor to #da3b01 in apiProperties.json

Changed the iconBrandColor value from #e0e0e0 to #da3b01 as requested in the PR review feedback. This aligns with the standard brand color requirement for Independent Publisher connectors.

DEmodoriGatsuO commented:
Hello @vmanoharas 

Thank you for the feedback. I've updated the `iconBrandColor` in the apiProperties.json file from "#e0e0e0" to "#da3b01" as requested. Please let me know if there are any other changes needed.

vmanoharas reviewed (approved):
Dear Partner

I hope you are doing well.
Congratulations, your pull request is approved. We will move forward with the certification process. We are preparing your connector for production deployment and will queue up for the next deployment schedule. Once your connector onboards the next deployment schedule, it starts to deploy your connector in our production environments that typically takes 3-4 weeks.
Please let us know if you have any questions.
Thank you very much for working with us.

vmanoharas commented:
Hello Partner,
I hope you are doing well.
Congratulations, the deployment of your connector and documentation page was already completed, 
Thank you very much for working with us.
